Add the 'estimate' button on transfer modal plus further UI fixes

antomor opened the pull request:
Add the estimation button on the transfer modal. We decided to use the on-chain estimation, which requires that the user signs the transaction.
Fix other UI issues.

## What

- Add the estimation button
- Show the rbtc/trif switch on the transfer modal used to select the token to use to pay for the fees
- Fix estimate loading on deploy modal
- Fix the loading status that was shared between the estimate and execute buttons
- Fix the return data slide button in the execute modal

## Why 

- In order to identify how much transactions cost, the estimate button is added in the transfer modal. It was the only modal not having such functionality.

mortelli commented on src/modals/Transfer.js:
```diff
@@ -103,6 +109,57 @@ function Transfer(props) {
         });
     }
+    
+      async function handleEstimateDeploySmartWalletButtonClick() {
```
should this be named "estimate transfer" instead?

antomor replied:
Yes, thanks for pointing it out

mortelli commented on src/Utils.js:
```diff
@@ -85,4 +87,43 @@ class Utils {
     }
 }
 
+const ESTIMATED_GAS_CORRECTION_FACTOR = 1.0;
+const INTERNAL_TRANSACTION_ESTIMATE_CORRECTION = 20000;
```
sorry, could you please explain the `20000`?

antomor replied:
The `estimateDestinationContractCallGas` method was extracted from the `ContractInteractor` class in `rif-relay-common` package. I reported the comments that were in the initial implementation. However, I guess that the value of 2000 was extracted after empirically testing different values.

mortelli replied:
thanks for adding the comments :+1:

mortelli commented on src/Utils.js:
```diff
@@ -85,4 +87,43 @@ class Utils {
+    const txDetailsClone = {
+        ...trxDetails
+    };
+    const internalCallCost = estimateDestinationContractCallGas(
+              relayClient.getEstimateGasParams(txDetailsClone)
+    );
+    txDetailsClone.gas = toHex(internalCallCost);
+    const tokenGas = (
+        await relayClient.estimateTokenTransferGas(txDetailsClone, process.env.REACT_APP_CONTRACTS_RELAY_WORKER)
+    ).toString();
+    txDetailsClone.tokenGas = tokenGas;
+    const maxPossibleGasValue = await relayClient.estimateMaxPossibleRelayGas(
```
i'm trying to understand what the difference is (conceptually) between the `estimateDestinationContractCallGas` result and the `estimateMaxPossibleRelayGas` result. i suppose the 2nd one is the tx "simulation", but why does it need data from the first call? is it just for the gas?

antomor replied:
Hoping that the comments may clarify, in my understanding the `estimateDestinationContractCallGas` estimate the gas consumption sent to the destination contract without considering the token payment, while the `estimateMaxPossibleRelayGas` include all the gas spent.

mortelli replied:
thanks for clarifying this, resolving

mortelli reviewed (commented):
i've left some comments below.

i tried manually testing this, and even though the estimation seems to work OK for the transfer modal, i am not able to execute it on the dapp. it's telling me that the relay verifier has rejected the call.

will provide more info as soon as i have it, not really sure what is going on here

mortelli reviewed (approved):
so as discussed, i decided to test this out with the regular branches instead of the revenue sharing ones, since we can move this code there later.

it seems to work OK, although i want to mention 2 things:
- for the transfer modal, if you click on estimate without filling in the other fields first, an error occurs and you won't be able to estimate again until refreshing the page
- since we don't have the revenue sharing code, i couldn't trigger a rejection from the server when submitting lower fees than what was estimated

antomor commented:
First of all, thanks for reviewing it.

> for the transfer modal, if you click on estimate without filling in the other fields first, an error occurs and you won't be able to estimate again until refreshing the page

Thanks for pointing it out, I think I fixed it.

> since we don't have the revenue sharing code, i couldn't trigger a rejection from the server when submitting lower fees than what was estimated

Yep, maybe you could check if the server logs something related to the gas estimation, but I think we added these logs there just on the demo branch.

mortelli reviewed (approved):
thanks :+1:
